src/brogue/Architect.c

```diff
@@ -3445,8 +3445,10 @@ void restoreMonster(creature *monst, short **mapToStairs, short **mapToPit) {
         } else {
             theMap = mapToStairs;
         }
+
+        pmap[*x][*y].flags &= ~HAS_MONSTER;
         if (theMap) {
-            turnCount = ((theMap[monst->xLoc][monst->yLoc] * monst->movementSpeed / 100) - monst->status[STATUS_ENTERS_LEVEL_IN]);
+            turnCount = (theMap[monst->xLoc][monst->yLoc] - (monst->status[STATUS_ENTERS_LEVEL_IN] * 100 / monst->movementSpeed));
```

Comment:
If a monster is 10 steps from the stairs, its STATUS_ENTERS_LEVEL_IN is 11. If you're off-level for 3 turns, STATUS_ENTERS_LEVEL_IN is 7 (probably unintentional but it works) and turnCount is 3, so it paths 3 steps from the monster. I think it works as intended.

The issue is with slow monsters, if yendor is 10 steps from the stairs its STATUS_ENTERS_LEVEL_IN is 21. If you're off-level for 8 turns that becomes 12, and turnCount is 10*2 - 12 = 8. So yendor moves 8 turns in the time it should have moved 4. Basically slow monsters take half as much time to move and camp the door for the other half of their turns, and I don't even think fast monsters even reach the door. The right turnCount would be 10 - 12/2 = 4.

The STATUS_ENTERS_LEVEL_IN quirk is because of a call to synchronizePlayerTimeState at the beginning of startLevel, which makes some parts of a turn get executed twice when a level is entered. I noticed this a while back because my 2 turn tele charm recharged instantly after using it on my stair turn, but I didn't want to remove it because I wasn't sure what it would break.

Reply:
Ok, I see, so STATUS_ENTERS_LEVEL_IN needs to be converted from a speed-adjusted value (which is decremented by 1 every turn) to a distance value, so it can be properly subtracted from the distance map. Makes sense! Adding a comment here summarising that would be great I think